src/migrate.c

```diff
@@ -0,0 +1,390 @@
+    if (e != 0) {
+        replyRaftError(req->ctx, e);
+        entryDetachRaftReq(rr, entry);
+        raft_entry_release(entry);
+        goto fail;
+    }
+
+    raft_entry_release(entry);
+
+    return REDISMODULE_OK;
+
+fail:
+    RaftReqFree(req);
+    return REDISMODULE_OK;
+}
+
+static void raftAppendRaftUnlockDeleteEntry(RedisRaftCtx *rr, RaftReq *req)
+{
+    raft_entry_resp_t response;
+
+    raft_entry_t *entry = RaftRedisLockKeysSerialize(req->r.migrate_keys.keys, req->r.migrate_keys.num_keys);
+    entry->id = rand();
+    entry->type = RAFT_LOGTYPE_DELETE_UNLOCK_KEYS;
+    entryAttachRaftReq(rr, entry, req);
+
+    int e = raft_recv_entry(rr->raft, entry, &response);
+    if (e != 0) {
+        RedisModule_ReplyWithError(req->ctx, "ERR Unable to unlock/delete migrated keys, try again");
+        replyRaftError(req->ctx, e);
+        entryDetachRaftReq(rr, entry);
+        raft_entry_release(entry);
+        goto error;
+    }
+
+    raft_entry_release(entry);
+
+    /* Unless applied by raft_apply_all() (and freed by it), the request
+     * is pending so we don't free it or unblock the client.
+     */
+    return;
+
+error:
+    RaftReqFree(req);
+}
+
+static void transferKeysResponse(redisAsyncContext *c, void *r, void *privdata)
+{
+    Connection *conn = privdata;
+    JoinLinkState *state = ConnGetPrivateData(conn);
+    RedisRaftCtx *rr = ConnGetRedisRaftCtx(conn);
+    RaftReq *req = state->req;
+
+    redisReply *reply = r;
+
+    if (!reply) {
+        ConnMarkDisconnected(conn);
+        RedisModule_ReplyWithError(req->ctx, "ERR: connection dropped impporting keys into remote cluster, try again");
+        RaftReqFree(req);
+    } else if (reply->type == REDIS_REPLY_ERROR) {
+        ConnAsyncTerminate(conn);
+        replyWithFormatErrorString(req->ctx, "RAFT.IMPORT failed: %.*s", (int) reply->len, reply->str);
+        RaftReqFree(req);
+    } else if (reply->type != REDIS_REPLY_STATUS || reply->len != 2 || strncmp(reply->str, "OK", 2)) {
+        ConnAsyncTerminate(conn);
+        /* FIXME: above should be changed to string eventually? */
+        replyWithFormatErrorString(req->ctx, "ERR: received unexpected response from remote cluster, type = %d (wanted %d), len = %ld, response = %.*s", reply->type, REDIS_REPLY_STATUS, reply->len, (int) reply->len, reply->str);
+        RaftReqFree(req);
+    } else {
+        /* SUCCESS */
+        ConnAsyncTerminate(conn);
+        raftAppendRaftUnlockDeleteEntry(rr, req);
+    }
+
+    redisAsyncDisconnect(c);
+}
+
+static void transferKeys(Connection *conn)
+{
+    RedisRaftCtx *rr = ConnGetRedisRaftCtx(conn);
+    JoinLinkState *state = ConnGetPrivateData(conn);
+    RaftReq *req = state->req;
+
+    /* Connection is not good?  Terminate and continue */
+    if (!ConnIsConnected(conn)) {
+        return;
+    }
+
+    ShardGroup * sg = getShardGroupById(rr, req->r.migrate_keys.shardGroupId);
+    if (sg == NULL) {
+        //FIXME: error
```

Comment:
Should we add at least an error log here?

Reply:
in looking at this code, it seems like it might be extraneous now, don't see sg referenced anywhere.

Reply:
removed all sg code